File: solutions/haskell/secret-handshake/1/src/SecretHandshake.hs
module SecretHandshake (handshake) where

handshake :: Int -> [String]
handshake n
  | n >= 16 = reverse $ handshake (n - 16)
  | otherwise = hs n [] ["wink", "double blink", "close your eyes", "jump"]

hs :: Int -> [String ] -> [String] -> [String]
hs 0 outlist _ = reverse outlist
hs n outlist (w:wx)
  | odd n = hs (div n 2) (w:outlist) wx
  | otherwise  = hs (div n 2) outlist wx
hs _ _ _ = error "This should not happen!"

File: solutions/python/secret-handshake/1/secret_handshake.py
def commands(binary_str):
    actions = ["wink", "double blink", "close your eyes", "jump"]
    result = []
    digits = list(reversed(binary_str))
    reverse = digits.pop() == '1'
    for index, digit in enumerate(digits):
        if digit == '1':
            result.append(actions[index])
    if reverse: 
        result.reverse()
    return result